### ISSUE TITLE
Tunnel spawn or notify?

### ISSUE THREAD

**Issue by H4mlet** [FaceDeer/deep_roads#1]
Minetest 0.4.16, player:set_physics_override({speed = 5}), full view range, wireframe... I've been looking for a tunnel (between -10 and -2300) for about 30mins with no luck, even teleporting where the console notified the action-emerges.

Is there a way to receive a sort of notice ("debug mode") telling you where a tunnel is being placed?
Something like "deep road placed @ <coords>" would be enough. :)

**Comment by FaceDeer** [comment]
Heh, haven't even really published this mod yet and already an issue is filed. :)

I'm planning to have the mod create its own special "log" text file that contains the locations of intersections as a sort of map for the server administrator, but for the time being you can get that stuff in the regular output by uncommenting the minetest.debug call in init.lua inside the "context:carve_intersection" loop. The "points" that it's being fed are the coordinates of the intersections, with an additional "val" property that is used to pseudorandomly generate the intersection's name.

I'm also thinking of adding a hook to allow a server administrator to _force_ an intersection to be generated at a particular location, if you want to have some predefined hubs to allow new players to get into the deep roads right away. Not in yet, obviously, but I can prioritize that if you think it'd be handy for your purposes.

I haven't done any serious work to tweak the density of the deep roads yet, so not surprised they're hard to find, during testing I've just stuck to an existing roadway and followed it wherever it went to find more. Before I release this mod I'll be needing to come up with a variety of presets for whatever playstyle people want - you could have rare, sparse intersections with kilometers between them or you could have a dense mesh of spaghetti lacing the underground, or anything in between that. Could even crank up the randomness and have some areas be dense and others be sparse. It will eventually be possible to have multiple separate deep road networks with different tunnel types and characteristics (eg, a dense mesh of narrow, shallow walking tunnels mixed with a sparse mesh of long-range rail tunnels that go deeper).

**Comment by H4mlet** [comment]
Hey there,
I've uncommented that debug call and eventually I've been able to find a large room with a mese lamp and a wooden sign; this is quite interesting for my game!

Shortly:
- I've developed a [dwarves](https://forum.minetest.net/viewtopic.php?f=9&t=20063&hilit=sidhe) mod that I could tweak to force the spawning in these underground roads/rooms.
- I've developed a humans mod that features a random string generator that follows some basic rules to grant pronounceable words, e.g. "Wotshal", "Sidzak", etc. no "RtGGkndd"-like garbage. If you're interested I'll be glad to give you that function. (constructed languages fan here)

To sum up, I'm very interested in this mod 'cause I think it could be a dwarf-fortress/moria-like "biome" for those dwarves that I'd like to add to my game.

On the issues side, (yeah I know, I can be a royal pain ;) ) I think that there's some glitch in that debug call, for it keeps printing the same coords over and over: https://gist.github.com/H4mlet/cfc387d29442ae337705e3478c7e3d6d

**Comment by H4mlet** [comment]
Hey,
haven't been able to check the latest changes - I'm improving my old mods :6 - but managed to clean up the random string function; I suggest you to try it on a "on rightclick" test-node, I've thoroughly checked it but you never know! Here it is: [random string function](https://gist.github.com/H4mlet/2f54680c4b4c2d28fb29c0c52e45f5bd)

**Comment by FaceDeer** [comment]
I've added your name generator as an optional predefine. There's now a "name_generator" property on the network_def table that takes a function with a single parameter, a floating point value from 0-1 that acts as the seed for generating a name and returns a string. The two test tunnel networks currently defined in init.lua each uses a different name generator, the shallow pedestrian network uses yours and the deeper rail tunnels use the two-random-word one. I'll give the default predefined generators better names later.

**Comment by H4mlet** [comment]
It seems that using a seed reduces the number of combinations, I've tried the modified function with a "on right click" test node which sends in chat the generated random string; instead of generating always new ones, it sticks to a couple or four. Maybe using os.clock() as random seed would be a better choice?

**Comment by FaceDeer** [comment]
I'll take a look at whether the seed isn't being used to generate entropy correctly, but it is necessary that the name function be able to deterministically generate a name from its input parameter. The mapgen function is run independently in each chunk that generates, so it uses a seed based in the location of the intersection to ensure that it "randomly" generates the same name each time it generates that particular intersection.

**Comment by H4mlet** [comment]
Ah! Got it, nevermind then.
